src/main/java/com/google/error_analyzer/backend/Pagination.java

```diff
@@ -0,0 +1,211 @@
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+
+@WebServlet("/pagination")
+public class Pagination extends HttpServlet {
+
+    public int recordsPerPage = 3;
+    // keep it a odd no so that there are equal pages in front and back 
+    public int noOfPages = 5;
+    public int extraPageInFrontAndBack = noOfPages / 2;
+    public String field = "name";
+    public String ERROR="errors";
+    public int noOfRecordsOnLastPage = recordsPerPage;
+    public int lastPage = Integer.MAX_VALUE;
+    public ArrayList < String > data = new ArrayList();
+    private static final Logger LOG = LogManager.getLogger(Pagination.class);
+    public RestHighLevelClient client = new RestHighLevelClient(RestClient.builder(new HttpHost("35.194.181.238", 9200, "http")));
+    public SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
+    // return object 
+    private class logOrErrorResponse {
+        boolean lastPage;
+        ArrayList < String > logOrError;
+        public logOrErrorResponse(ArrayList < String > logOrError, boolean lastPage) {
+            this.logOrError = logOrError;
+            this.lastPage = lastPage;
+        }
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        int page = Integer.parseInt(request.getParameter("requestedPage"));
+        String fileName = request.getParameter("fileName");
+        String fileType = request.getParameter("fileType");
+        String next = request.getParameter("next");
+
+        SearchErrors searchErrors = new SearchErrors();
+        HashMap < String, String > search = searchErrors.getSearchedErrors();
+        if (page <= 0)
+            page = 1;
+        // if file field is empty the user haven't choosen anyfile so return empty object
+        if (fileName.length() == 0) {
+            returnEmptyObject(response);
+            return;
+        }
+
+        // if fileName is not empty fetch the data from database and return response
+        fetchAndReturnResponse(page, fileName, fileType, next, search, response);
+
+    }
+
+    //  if user is asking for 1st page return that page and make a window of fully finished and ready to use data for that file
+    // if user is not on 1st page but someother page just maintain the already created window of pages
+
+    public void fetchAndReturnResponse(int page, String fileName, String fileType, String next, HashMap < String, String > search, HttpServletResponse response) throws IOException {
+        if (page == 1) {
+            // the user won't be knowing the name of error file generated corresponding to his log file so we need to change fileName depending on type of file
+            if (fileType.equals(ERROR))
```

Comment:
Parenthesis

Reply:
If page > 1, how will it know the fileName? Also why is there a separate fileName for errors?

Reply:
whenever user change file we start paginated view with page 1 so when page !=1 we can deliver the already fetched data from maintained window .
errors are kept separately in other index after being identified at analyzer layer